### PR TITLE
cgmgr: create cgroups for systemd cgroup driver for dropped infra pods

### PR DESCRIPTION
**Pull request by haircommander**
The history here is a bit convoluted. Originally, runc created the cgroup for the infra container.
    cAdvisor was built to assume the cgroup for the infra container would be created, and it uses
    this to find the network metrics for the pod. When we dropped the infra container, cri-o needed to make this
    cgroup so cAdvisor could still find the network metrics.
    
    However, systemd didn't like the way we did it, and would remove the cgroup mid pod creation, which was fixed in
    https://github.com/cri-o/cri-o/pull/6196. This actually caused the cgroup to not be created at all, which then caused
    the networking metrics to not be gathered at all.
    
    Thus, we do need to create a cgroup underneath the systemd cgroup. Attempt to use a slice for this, as systemd won't require a
    process be underneath it.

#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
reverts https://github.com/cri-o/cri-o/pull/6196 
fixes https://github.com/cri-o/cri-o/issues/6657 
#### Does this PR introduce a user-facing change?

```release-note
Fix a bug where network metrics collection is broken with systemd cgroup driver and dropped infra containers.
```

### CODE REVIEW

**Comment by haircommander** [comment]
/hold

I think there are some cgroupfs issues here and I haven't tested thoroughly with cgroupv2 yet either

File: internal/config/cgmgr/cgmgr.go
Comment (kolyshkin) [inline internal/config/cgmgr/cgmgr.go:77]: typo: `s/CreateSandboxCgroup/RemoveSandboxCgroup/`.

Perhaps we should use a linter to catch this.  Revive would do (alas enabling it for the whole repo will cause tons of warnings)

File: internal/config/cgmgr/cgmgr.go
Comment (kolyshkin) [inline internal/config/cgmgr/cgmgr.go:181]: Well this does not make much sense to me.

1. Since you haven't specified `UseSystemd: true`, you are using fs manager. It just creates a bunch of directories.
2. It's a scope, so its lifetime is bound to the existence of at least one process in the scope. As we don't have any processes, systemd may just remove this.
3. I am not sure that having a `name=systemd` controller in there helps with (2), but overall this smells bad -- we create something that looks very much like what systemd would create, but systemd knows nothing about it.

Let's step back. Here is my limited understanding of the problem.

We have a pod (which is a scope in systemd terms), and a bunch of containers in it. We used to have a dummy infra container running a single pause process in order for cadvisor to identify the pod and get its stats. Now we don't have an infra container, and other containers may come and go, so there's no container that will always exist in the pod, and as a result cadvisor loses the pod (stops reporting its stats).

If what's written above is correct, than this is a problem with the approach of referring to a pod via a particular container which is in it. Containers can come and go,  and a pod should be identified by a pod cgroup path or something like that, not via any container.

**Review by kolyshkin — CHANGES_REQUESTED** [review]

File: internal/config/cgmgr/cgmgr.go
Comment (haircommander) [reply internal/config/cgmgr/cgmgr.go:181]: > so there's no container that will always exist in the pod, and as a result cadvisor loses the pod (stops reporting its stats)

Sort of. cadvisor assumes there's an infra container and an associated cgroup. the only way it registers network metrics is if there's a cgroup associated with the pod container. We can trick it to gather network metrics for the pod by pretending like there's the infra container by creating this dummy cgroup. I agree it smells bad--I've been wanting to change this interation for years with CRI stats KEP, but it's been delayed a number of times.

addressing the points more directly:
> Since you haven't specified UseSystemd: true, you are using fs manager. It just creates a bunch of directories.

yes this is the point. libcontainer will create the empty cgroup, and cadvisor (Which as inotify on the cgroup hierarchy) will catch its creation, see the pod ID, and then ask cri-o about that pod

> It's a scope, so its lifetime is bound to the existence of at least one process in the scope. As we don't have any processes, systemd may just remove this.

I was wondering whether this will happen. I believe this is the cause of the issue requiring https://github.com/cri-o/cri-o/pull/6196. I am not sure if this PR will help that or not, I am hoping it wil..

> I am not sure that having a name=systemd controller in there helps with (2), but overall this smells bad -- we create something that looks very much like what systemd would create, but systemd knows nothing about it.

agreed, though I'm not sure we can achieve anything better in the meantime. Folks need their network metrics, and we want systemd to not remove the cgroup. if we can achieve both of those internally in cri-o, it will be worth it for now, while we work on fixing it more legitimately

**Comment by haircommander** [comment]
after a conversation with @kolyshkin offline, we opted to have crio create a slice for the infra container, as systemd won't require a process be underneath it.

File: test/pod.bats
Comment (kolyshkin) [inline test/pod.bats:285]: nit: this comment is now misplaced

File: internal/config/cgmgr/systemd.go
Comment (haircommander) [reply internal/config/cgmgr/systemd.go:236]: we basically do this now by having two versions of this function. that's the pattern we've adopted for branching between the two

Comment (kolyshkin) [reply internal/config/cgmgr/systemd.go:236]: Ah, sorry, this *IS* systemd cgroup manager, so it's always `true`.

File: test/cgroups.bats
Comment (kolyshkin) [inline test/cgroups.bats:94]: Offtopic, but that reminds I recently learned that this form won't work, unless this is the last statement in a function. I'll send a fix.

**Comment by kolyshkin** [comment]
So, I would like to have it tested before the merge, to make sure
 - the issue (#6657) is solved
 - the older bug (#6349) is not reintroduced

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by haircommander** [comment]
/hold

to double check 

>     * the issue ([Networking metrics missing with cri-o 1.26 #6657](https://github.com/cri-o/cri-o/issues/6657)) is solved
> 
>     * the older bug ([1.24.3 leaks pidns with cgroup v2 #6349](https://github.com/cri-o/cri-o/issues/6349)) is not reintroduced

**Comment by haircommander** [comment]
/hold cancel

* the issue (https://github.com/cri-o/cri-o/issues/6657) is solved 

I now see the network metrics

* the older bug ( https://github.com/cri-o/cri-o/issues/6349) is not reintroduced

I don't see a pid cgroup being created. I think using libcontainer has fixed this

I also have monitored the cgroup for a bit and i don't see systemd cleaning it up. I think we're all set

**Review by kolyshkin — APPROVED** [review]
/lgtm

**Pull request merged**
